Hidden members should not be shown in QuickInfo

forki opened the pull request:
this fixes #50

I added a testcase and I used `CheckFSharpAttributesForUnseen` to ask for the `hidden` attribute.

I had to split `CheckFSharpAttributesForUnseen` since `NicePrint.fs` has a mode `denv.showObsoleteMembers` which allows to show obsolete members but `CheckFSharpAttributesForUnseen` checked for hidden and obsolote.

dsyme commented:
Great!  LGTM. Especially good to see the Language Service unit test added.

dsyme commented on src/fsharp/NicePrint.fs:
```diff
@@ -1505,6 +1505,7 @@ module private TastDefinitionPrinting =
                                       not (isInterfaceTy denv.g oty)
                                   | [] -> true)
               |> List.filter (fun v -> denv.showObsoleteMembers || not (HasFSharpAttribute denv.g denv.g.attrib_SystemObsolete v.Attribs))
+              |> List.filter (fun v -> not (Infos.AttributeChecking.CheckFSharpAttributesForHidden denv.g v.Attribs))
```
The NicePrint.fs code is used to show the signature of things when emitting signature files using the "-sig:file" option. So I think you need to show obsolete, hidden and experimental members in that case, because the signature file should be "logically complete".  

For this, you should add a flag "showHiddenMembers" to DisplayEnv with default value "false", and set this to "true" in BuildInitialDisplayEnvForDocGeneration.

I actually think "showObsoleteMembers" should match this and be "false" by default, but set to "true" in BuildInitialDisplayEnvForDocGeneration.  I don't think we want obsolete members shown in class/module tooltips.

Finally, rename BuildInitialDisplayEnvForDocGeneration to BuildInitialDisplayEnvForSigFileGeneration.

dsyme commented:
@latkin @KevinRansom  - I went looking for an "under review" label to label the pull request with - I kind of like it if all issues and pull requetss get a label of some kind - if you agree could you create a label for this?  thanks!

forki commented:
yep seems you are correct. obsolete methods are still shown in QuickTips but not in IntelliSense.

![image](https://cloud.githubusercontent.com/assets/57396/5835278/dbbbd3ba-a166-11e4-8dbd-f33d56606a99.png)

![image](https://cloud.githubusercontent.com/assets/57396/5835282/f0859e16-a166-11e4-9aa6-9138a4078767.png)

I'll append a fix.

forki commented:
I added a test for the obsolete case and added the changes you suggested.

But now we have a new issue: the test is still red.
On my local repo I added a bit of printf debugging and saw that the Obsolete-attribute is not found in https://github.com/Microsoft/visualfsharp/pull/73/files#diff-4a6abf380cb46c97f62d8f796adeefdfR1507 since v.Attribs is empty. 

Any ideas?

dsyme commented:
I took a look and didn't spot any problem with the code.  Do both tests fail or just the "obsolete" one?  Is v.Attribs empty for both Print1 and Print2?

forki commented:
only the obsolete test fails. Both properties (Print1 and Print2) have 0 attribs. 

If I add the HiddenAttribute to Print1 in the obsolete test, I get 1 attribute (Obsolete attr is still missing) and Print1 is no longer displayed.

So somehow it doesn't find the ObsoleteAttribute, but the HiddenAttribute works.

dsyme commented:
You may have to open "System" or use "System.Obsolete"

forki commented:
scratch that. I cleaned my repo and rebuilt everything from scratch.
All green. :shipit: ;-)

dsyme commented:
OK, great, then this looks good to go.

forki commented:
just for my understanding:

the attribute wasn't found since the code didn't compile.
The QuickInfo still kind-of worked on the uncompilable code. It just processed everything without the attribute.

Is that correct?

forki commented:
Yes! First accepted bug fix.

![muscle](https://cloud.githubusercontent.com/assets/57396/5853608/5c7593b0-a225-11e4-9e92-95b7006d0e99.gif)
